core/templates/dev/head/domain/question/QuestionPlayerBackendAPIServiceSpec.js

```diff
@@ -0,0 +1,186 @@
+            },
+            id: 'TextInput'
+          },
+          param_changes: []
+        },
+        language_code: 'en',
+        version: 1
+      }],
+      next_start_cursor: null
+    };
+  }));
+
+  afterEach(function() {
+    $httpBackend.verifyNoOutstandingExpectation();
+    $httpBackend.verifyNoOutstandingRequest();
+  });
+
+  it('should successfully fetch questions from the backend', function() {
+    var successHandler = jasmine.createSpy('success');
+    var failHandler = jasmine.createSpy('fail');
+
+    $httpBackend.expect(
+      'GET', '/question_player_handler?skill_ids=1&question_count=1&' +
+      'start_cursor=').respond(
+      sampleDataResults);
+    QuestionPlayerBackendApiService.fetchQuestions(
+      '1', '1').then(successHandler, failHandler);
+    $httpBackend.flush();
+
+    expect(successHandler).toHaveBeenCalledWith(
+      sampleDataResults.question_dicts);
+    expect(failHandler).not.toHaveBeenCalled();
+  });
+
+  it('should successfully fetch questions from the backend with given cursor',
+    function() {
+      var successHandler = jasmine.createSpy('success');
+      var failHandler = jasmine.createSpy('fail');
+
+      var sampleDataResultsWithCursor = sampleDataResults;
+      sampleDataResultsWithCursor.next_start_cursor = '1';
+
+      $httpBackend.expect(
+        'GET', '/question_player_handler?skill_ids=1&question_count=1&' +
+        'start_cursor=').respond(
+        sampleDataResultsWithCursor);
+      QuestionPlayerBackendApiService.fetchQuestions(
+        '1', '1').then(successHandler, failHandler);
+      $httpBackend.flush();
+
+      expect(successHandler).toHaveBeenCalledWith(
+        sampleDataResultsWithCursor.question_dicts);
+      expect(failHandler).not.toHaveBeenCalled();
+
+      $httpBackend.expect(
+        'GET', '/question_player_handler?skill_ids=1&question_count=1&' +
+        'start_cursor=1').respond(
+        sampleDataResults);
+
+      QuestionPlayerBackendApiService.fetchQuestions(
+        '1', '1').then(successHandler, failHandler);
+      $httpBackend.flush();
+
+      expect(successHandler).toHaveBeenCalledWith(
+        sampleDataResults.question_dicts);
+      expect(failHandler).not.toHaveBeenCalled();
+    });
+
+  it('should use the fail handler if the backend request failed',
+    function() {
+      var successHandler = jasmine.createSpy('success');
+      var failHandler = jasmine.createSpy('fail');
+
+      $httpBackend.expect(
+        'GET', '/question_player_handler?skill_ids=1&question_count=1&' +
+        'start_cursor=').respond(
+        500, 'Error loading questions.');
+      QuestionPlayerBackendApiService.fetchQuestions(
+        '1', '1').then(successHandler, failHandler);
+      $httpBackend.flush();
+
+      expect(successHandler).not.toHaveBeenCalled();
+      expect(failHandler).toHaveBeenCalledWith(
+        'Error loading questions.');
+    }
+  );
+
+  it('should use the fail handler if question count is in invalid format',
```

Comment:
Add more failure cases to these tests (cf. previous comment).

Reply:
added